File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.py
def chk_level(diffs, times, limit, level):
    tot_times = 0
    n = len(diffs)
    
    for i in range(n):
        cur_diff = diffs[i]
        cur_time = times[i]
        prev_time = times[i - 1]
        
        if cur_diff <= level:
            tot_times += cur_time
        else:
            tot_times += (prev_time * (cur_diff - level)) + (cur_time * (cur_diff - level)) + cur_time
            
        if tot_times > limit:
            return False
    
    return True


def solution(diffs, times, limit):
    answer = 0
    
    max_num = max(diffs)
    min_num = 1
    
    while min_num <= max_num:
        mid_num = (min_num + max_num) // 2
        
        res = chk_level(diffs, times, limit, mid_num)
        
        if res:
            max_num = mid_num - 1
        else:
            min_num = mid_num + 1
    
    return min_num

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js

function chk_time(diffs, times, limit, level) {
    let totalTime = 0
    
    for(let i in diffs) {
        let cur_diff = diffs[i]
        let cur_time = times[i]
        
        if(cur_diff <= level) totalTime += cur_time
        else {
            let prev_time = times[i-1]
            totalTime += (prev_time + cur_time) * (cur_diff - level) + cur_time
        }
        
        if (totalTime > limit) return false
    }
    return true
}

function solution(diffs, times, limit) {
    let left = 1
    let right = [...diffs].sort((a, b) => b - a)[0]
    
    while(left <= right) {
        let mid = Math.floor((left + right) / 2)
        let res = chk_time(diffs, times, limit, mid)
        
        if(res) right = mid - 1
        else left = mid + 1
        
    }
    return left;
}